src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
+    const { address } = walletTools.scriptPubkeyToAddress({
+      scriptPubkey,
+      format
+    })
+    taskCache.addressSubscribeCache.set(address, { format, branch })
+    taskCache.addressWatching = false
+  }
 }
 
-interface ProcessAddressTxsArgs extends FormatArgs {
+interface ProcessAddressTxsArgs
+  extends AddressTransactionCacheState,
+    CommonArgs {
   address: string
-  page?: number
-  networkQueryVal?: number
 }
 
+type addressResponse = IAccountDetailsBasic &
+  ITransactionDetailsPaginationResponse
+
 const processAddressTransactions = async (
   args: ProcessAddressTxsArgs
-): Promise<void> => {
-  const { address, page = 1, processor, walletTools, blockBook } = args
+): Promise<WsTask<addressResponse>> => {
+  const {
+    address,
+    page = 1,
+    networkQueryVal,
+    processor,
+    walletTools,
+    path,
+    taskCache
+  } = args
+  const transactionsCache = taskCache.transactionsCache
 
   const scriptPubkey = walletTools.addressToScriptPubkey(address)
   const addressData = await processor.fetchAddressByScriptPubkey(scriptPubkey)
-  const networkQueryVal = args.networkQueryVal ?? addressData?.networkQueryVal
-  const {
-    transactions = [],
-    txs,
-    unconfirmedTxs,
-    totalPages
-  } = await blockBook.fetchAddress(address, {
-    details: 'txs',
-    from: networkQueryVal,
-    perPage: BLOCKBOOK_TXS_PER_PAGE,
-    page
-  })
 
-  // If address is used and previously not marked as used, mark as used.
-  const used = txs > 0 || unconfirmedTxs > 0
-  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
-  if (used && !addressData?.used && page === 1) {
-    await processor.updateAddressByScriptPubkey(scriptPubkey, {
-      used
-    })
-    await setLookAhead(args)
-  }
+  const deferred = new Deferred<addressResponse>()
+  deferred.promise
+    .then(async (value: addressResponse) => {
+      const { transactions = [], txs, unconfirmedTxs, totalPages } = value
 
-  for (const rawTx of transactions) {
-    const tx = processRawTx({ ...args, tx: rawTx })
-    await processor.saveTransaction(tx)
-  }
+      // If address is used and previously not marked as used, mark as used.
+      const used = txs > 0 || unconfirmedTxs > 0
+      if (used && !(addressData?.used ?? false) && page === 1) {
+        await processor.updateAddressByScriptPubkey(scriptPubkey, {
+          used
+        })
+        await setLookAhead({ ...args, ...path })
+      }
+
+      for (const rawTx of transactions) {
+        const tx = processRawTx({ ...args, tx: rawTx })
+        await processor.saveTransaction(tx)
+      }
+
+      if (page < totalPages) {
+        // Add the address back to the cache, incrementing the page
+        transactionsCache.set(address, {
+          path,
+          networkQueryVal,
+          fetching: false,
+          page: page + 1
+        })
+      } else {
+        // Callback for when an address has been fully processed
+        args.onAddressChecked()
```

Comment:
Call this when the addresses has run through all cache tasks, e.g. when it is no longer in the cache.